### PR TITLE
Adds a colorblind accessability testing tool

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

I keep finding myself worrying about if things I create will be parsable for colorblind people. 
So I've made a debug tool for approximating different extreme forms of colorblindness.

It's very very much a hack. 
We can't do the proper correction required to actually deal directly with long medium and short wavelengths of light, so we need to rely on approximations. 
Part of that means say, bright things being brighter then they ought to be. 
S not how people actually experience things, but it's not something we can do anything about in byond.

Anyway uh, it works by taking color matrixes, and using the plane master grouping system floyd added to apply them to most all parts of the game you would want to color correct.

There's some slight fragility here, but I couldn't think of a better way of handling it.

We also need to deal with planes that have BLEND_MULTIPLY as their blendmode, since that fucks up the filter. I've come up with a hack for it, since I wanted to avoid breaking anything.

Oh and since I want it to apply to huds too I added plane masters to represent them. I think that's about it.

image

## Changelog

:cl: LemonInTheDark, MrStonedOne
code: Adds a color blindness accessibility testing tool. Use it for testing out your ui changes
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
we have an accessibility tag that doesn't have a changelog icon, but it does affect GBP

**Comment by LemonInTheDark** [comment]
I am more powerful then you could possibly imagine

**Comment by JohnFulpWillard** [comment]
^ doesn't have enough GBP for embed

**Comment by kawoppi** [comment]
Wouldn't it be easier to do this with an external program? [Color Oracle](http://colororacle.org/index.html) can do this to the entire screen, but unfortunately it goes back to normal when you click somewhere. [Colour Simulations](http://www.coloursimulations.com/home.html) can do it "live" as a screen overlay when you use it in fullscreen. So you can play the game with it enabled.

**Comment by LemonInTheDark** [comment]
Color oracle is bad and its source code makes me sad. Didn't know color simulations existed, I think it's at least partially useful to have a "use this for a quick check" tool, but maybe that's just the bias talking.

**Comment by MrStonedOne** [comment]
> Wouldn't it be easier to do this with an external program? [Color Oracle](http://colororacle.org/index.html) can do this to the entire screen, but unfortunately it goes back to normal when you click somewhere. [Colour Simulations](http://www.coloursimulations.com/home.html) can do it "live" as a screen overlay when you use it in fullscreen. So you can play the game with it enabled.

When I was helping it seemed that some of them have the same matrixes with the same inaccuracies (the key thing is that the "real" matrix requires the use of two matrixes depending on the value of each rgb pixel, its not linear so you can't use a linear transform which knocks out most hardware accelerated ways of applying color transforms via matrixes to screen output)

Also that doesn't help the end goal, which is make it easy for spriters to gauge the colorblind accessibility of a given thing.

Every spriter has byond and the tg codebase installed, its unlikely this would be true of an external application unless it was a portable exe that got packaged into the tools folder of the main repo. Any further inertia or friction would severely drop actual usage metrics.

**Comment by LemonInTheDark** [comment]
To be clear, we can't use the real way either, but people are historically quite bad at doing it proper

File: code/modules/admin/verbs/color_blind_test.dm [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		return
+
+	var/datum/hud/our_hud = ui.user?.hud_used
+	if(!our_hud) // Nothing to act on
+		return
+
+	switch(action)
+		if("set_matrix")
+			set_selected_type(params["name"], our_hud)
+			. = TRUE
```
Comment (Mothblocks) [inline code/modules/admin/verbs/color_blind_test.dm:52]:
```suggestion
			return TRUE
```

File: code/modules/admin/verbs/color_blind_test.dm [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		if("clear_matrix")
+			set_selected_type("", our_hud)
+			. = TRUE
```
Comment (Mothblocks) [inline code/modules/admin/verbs/color_blind_test.dm:55]:
```suggestion
			return TRUE
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Pull request merged**